python/hail2/dataset.py

```diff
@@ -82,241 +344,354 @@ def genotype_schema(self):
             self._genotype_schema = Type._from_java(self._jvds.genotypeSignature())
         return self._genotype_schema
 
-
-class AggregatedVariantDataset(DatasetTemplate):
-    def __init__(self, hc, jvds, scope):
-        super(AggregatedVariantDataset, self).__init__(hc, jvds)
-        self.__setattr__("variants", convert_column(Column("variants", TAggregable(self.variant_schema), scope=scope)))
-        self.__setattr__("samples", convert_column(Column("samples", TAggregable(self.sample_schema), scope=scope)))
-
     @handle_py4j
-    @typecheck_method(exprs=tupleof(Column))
-    def query_samples_typed(self, *exprs):
-        if len(exprs) > 1:
-            exprs = [to_expr(e) for e in exprs]
-            result_list = self._jvds.querySamples(jarray(Env.jvm().java.lang.String, exprs))
-            ptypes = [Type._from_java(x._2()) for x in result_list]
-            annotations = [ptypes[i]._convert_to_py(result_list[i]._1()) for i in xrange(len(ptypes))]
-            return annotations, ptypes
-        else:
-            result = self._jvds.querySamples(to_expr(exprs[0]))
-            t = Type._from_java(result._2())
-            return t._convert_to_py(result._1()), t
+    def annotate_globals(self, **named_exprs):
+        exprs = []
+        named_exprs = {k: to_expr(v) for k, v in named_exprs.items()}
+        for k, v in named_exprs.items():
+            analyze(v, self, set(), set(), {'globals'}, False)
+            exprs.append('global.`{k}` = {v}'.format(k=k, v=v._ast.to_hql()))
+
+            if k in self._reserved:
+                msg = 'name collision with reserved namespace: {}'.format(k)
+                error('Analysis exception: {}'.format(msg))
+                raise ExpressionException(msg)
+            if k in self._all_fields and not k in [f.name for f in self.variant_schema.fields]:
+                msg = 'name collision with outside namespace: {}'.format(k)
+                error('Analysis exception: {}'.format(msg))
+                raise ExpressionException(msg)
+        base, cleanup = self._process_joins(*named_exprs.values())
+        m = VariantDataset(self._hc, base._jvds.annotateGlobalExpr(",\n".join(exprs)))
+        return cleanup(m)
 
     @handle_py4j
-    @typecheck_method(exprs=tupleof(Column))
-    def query_samples(self, *exprs):
-        r, t = self.query_samples_typed(*exprs)
-        return r
+    def annotate_rows(self, **named_exprs):
+        exprs = []
```

Comment:
The body of this method is broadly duplicated throughout this file.  I think you can have a single method that contains it that processes the exprs.  That method will throw the error exceptions, so you'll get one level in the stack track with an evocative name like "check_exprs".  What do you think?